File: api/src/main/java/mb/pie/api/stamp/resource/Modified.java
package mb.pie.api.stamp.resource;

import mb.resource.ReadableResource;
import mb.resource.fs.FSResource;
import mb.resource.fs.match.ResourceMatcher;
import mb.resource.fs.walk.ResourceWalker;
import org.checkerframework.checker.nullness.qual.Nullable;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.util.stream.Stream;

class Modified {
    static long modified(ReadableResource resource) throws IOException {
        if(!resource.exists()) {
            return getMaximum();
        } else {
            return modifiedResource(resource);
        }
    }

    static long modified(FSResource resource, @Nullable ResourceMatcher matcher) throws IOException {
        if(!resource.exists()) {
            return getMaximum();
        } else if(resource.isFile()) {
            return modifiedResource(resource);
        } else if(resource.isDirectory()) {
            return modifiedDir(resource, matcher);
        }
        return getMinimum();
    }

    static long modifiedRec(FSResource resource, @Nullable ResourceWalker walker, @Nullable ResourceMatcher matcher) throws IOException {
        if(!resource.exists()) {
            return getMaximum();
        } else if(resource.isFile()) {
            return modifiedResource(resource);
        } else if(resource.isDirectory()) {
            return modifiedDirRec(resource, walker, matcher);
        }
        return getMinimum();
    }


    private static long modifiedResource(ReadableResource resource) throws IOException {
        return resource.getLastModifiedTime().toEpochMilli();
    }

    private static long modifiedDir(FSResource dir, @Nullable ResourceMatcher matcher) throws IOException {
        final long[] lastModified = {getMinimum()}; // Use array to allow access to non-final variable in closure.
        final boolean useMatcher = matcher != null;
        try(final Stream<? extends FSResource> stream = useMatcher ? dir.list(matcher) : dir.list()) {
            stream.forEach((resource) -> {
                try {
                    final long modified = modifiedResource(resource);
                    lastModified[0] = Math.max(lastModified[0], modified);
                } catch(IOException e) {
                    throw new UncheckedIOException(e);
                }
            });
        } catch(UncheckedIOException e) {
            throw e.getCause();
        }
        return lastModified[0];
    }

    private static long modifiedDirRec(FSResource dir, @Nullable ResourceWalker walker, @Nullable ResourceMatcher matcher) throws IOException {
        final long[] lastModified = {getMinimum()}; // Use array to allow access to non-final variable in closure.
        final boolean useWalker = walker != null && matcher != null;
        try(final Stream<? extends FSResource> stream = useWalker ? dir.walk(walker, matcher) : dir.walk()) {
            stream.forEach((resource) -> {
                try {
                    final long modified = modifiedResource(resource);
                    lastModified[0] = Math.max(lastModified[0], modified);
                } catch(IOException e) {
                    throw new UncheckedIOException(e);
                }
            });
        } catch(UncheckedIOException e) {
            throw e.getCause();
        }
        return lastModified[0];
    }

    private static long getMinimum() {
        return Long.MIN_VALUE;
    }

    private static long getMaximum() {
        return Long.MAX_VALUE;
    }
}

File: api/src/main/kotlin/mb/pie/api/fs/stamp/Modified.kt
package mb.pie.api.fs.stamp

import mb.fs.api.node.*
import mb.fs.api.node.match.FSNodeMatcher
import mb.fs.api.node.walk.FSNodeWalker
import mb.pie.api.fs.FileSystemResource

interface ModifiedResourceStamperTrait : FileSystemStamper {
  val unknown get() = Long.MIN_VALUE

  fun modified(node: FSNode, matcher: FSNodeMatcher?): Long {
    if(node.isDirectory) return modifiedDir(node, matcher)
    if(node.isFile) return node.lastModifiedTime.toEpochMilli()
    return unknown
  }

  fun modifiedRec(node: FSNode, walker: FSNodeWalker?, matcher: FSNodeMatcher?): Long {
    if(node.isDirectory) return modifiedDirRec(node, walker, matcher)
    if(node.isFile) return node.lastModifiedTime.toEpochMilli()
    return unknown
  }

  fun modifiedDir(dir: FSNode, matcher: FSNodeMatcher?): Long {
    if(matcher == null) return dir.lastModifiedTime.toEpochMilli()
    var lastModified = unknown
    dir.list(matcher).use { stream ->
      for(subPath in stream) {
        val modified = subPath.lastModifiedTime.toEpochMilli()
        lastModified = Math.max(lastModified, modified)
      }
    }
    return lastModified
  }

  fun modifiedDirRec(dir: FSNode, walker: FSNodeWalker?, matcher: FSNodeMatcher?): Long {
    var lastModified = unknown
    if(walker == null || matcher == null) {
      dir.walk()
    } else {
      dir.walk(walker, matcher)
    }.use { stream ->
      for(subPath in stream) {
        val modified = subPath.lastModifiedTime.toEpochMilli()
        lastModified = Math.max(lastModified, modified)
      }
    }
    return lastModified
  }
}

data class ModifiedResourceStamper @JvmOverloads constructor(
  private val matcher: FSNodeMatcher? = null
) : ModifiedResourceStamperTrait {
  override fun stamp(resource: FileSystemResource): FileSystemStamp {
    val node = resource.node
    if(!node.exists()) {
      return ValueResourceStamp(null, this)
    }
    val modified = modified(node, matcher)
    return ValueResourceStamp(modified, this)
  }

  override fun toString(): String {
    return "Modified($matcher)"
  }
}

data class RecModifiedResourceStamper @JvmOverloads constructor(
  private val walker: FSNodeWalker? = null,
  private val matcher: FSNodeMatcher? = null
) : ModifiedResourceStamperTrait {
  override fun stamp(resource: FileSystemResource): FileSystemStamp {
    val node = resource.node
    if(!node.exists()) {
      return ValueResourceStamp(null, this)
    }
    val modified = modifiedRec(node, walker, matcher)
    return ValueResourceStamp(modified, this)
  }

  override fun toString(): String {
    return "RecModified($walker)"
  }
}
